src/os/bluestore/BlueStore.cc

```diff
@@ -11903,13 +12098,32 @@ void BlueStore::_read_cache(
 int BlueStore::_prepare_read_ioc(
   blobs2read_t& blobs2read,
   vector<bufferlist>* compressed_blob_bls,
-  IOContext* ioc)
-{
+  IOContext* ioc,
+  span_stat_t* res_span_stat)
```

Comment:
`_prepare_read_ioc` is modified to provide scanning for optimization.
We should have a separate function that will scan object and decide action.
For example it might happen that in object we have 3 chunks for total of 4MB,
and in attempt to defragment it we allocated 3 chunks.

Reply:
The big reason I think we want to piggyback on deep-scrub is because we're already reading the content of the object to do the scrub, so it's the perfect time to make decisions about de-fragmentation and compression as the most expensive part of this can easily be reading the fragmented extents.  I don't think we should scan objects as a separate loop unless we have a very good reason.  It's much better in my mind to opportunistically defrag/compress when we're already done the expensive fragmented read.

Reply:
@markhpc
I (now) agree that performing optimization to object as an extra step during deep-scrub is good.
"I don't think we should scan objects as a separate loop unless we have a very good reason."
The code here is scanning the object - it's just piggybacking `_read_cache` to do it.
There is no change to behaviour of `_read_cache`. 
The filling of `span_stat_t` could be done in separate loop at a cost of a loop.
Similarly, we now  load shared blobs in `_prepare_read_ioc` - useless for reading process.
That would made optimization sense, if significant part of processing was reused - but its only the loop. 

@gardran My opinion is that whenever we make a new feature we should make it as detached from existing code as possible to minimize accidental bug injection. Withing reason of course...

Reply:
>@gardran My opinion is that whenever we make a new feature we should make it as detached from existing code as possible to minimize accidental bug injection. Within reason of course...

Generally I agree with this point when we're talking about a severe intervention into the execution path: changing run flow, having new algorithms, etc... 
Instead I just perform stats collection (by using simple arithmetic ops) here. So apart from a trivial null/invalid pointer dereferencing this shouldn't introduce any issues to existing functionality. Moreover if refactoring is disabled this function collects stats into a dummy local var which is an absolutely trivial operation.
So IMO that sort of changes (in this specific function) is quite safe and doesn't worth the separation - basic code review/testing should reveal any potential issues if any.